# 第一题：使用sklearn的GaussianNB、BernoulliNB、MultinomialNB完成spambase垃圾邮件分类任务

实验内容：
1. 使用GaussianNB、BernoulliNB、MultinomialNB完成spambase邮件分类
2. 计算各自十折交叉验证的精度、查准率、查全率、F1值
3. 根据精度、查准率、查全率、F1值的实际意义以及四个值的对比阐述三个算法在spambase中的表现对比

# 1. 读取数据集

In [1]:
import numpy as np
spambase = np.loadtxt('data/spambase/spambase.data', delimiter = ",")
spamx = spambase[:, :57]
spamy = spambase[:, 57]

# 2. 导入模型

In [2]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import cross_val_predict

# 3. 计算十折交叉验证下，GaussianNB、BernoulliNB、MultinomialNB的精度、查准率、查全率、F1值

In [8]:
# YOUR CODE HERE
model1 = GaussianNB()
prediction1 = cross_val_predict(model1,spamx, spamy, cv = 10)

model2 = BernoulliNB()
prediction2 = cross_val_predict(model2, spamx, spamy, cv = 10)

model3 = MultinomialNB()
prediction3 = cross_val_predict(model3, spamx, spamy, cv = 10)



In [10]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
import pandas as pd


acc_s = []
pre_s = []
rec_s = []
f1_s = []

acc_s.append(accuracy_score(spamy, prediction1))
pre_s.append(precision_score(spamy, prediction1))
rec_s.append(recall_score(spamy, prediction1))
f1_s.append(f1_score(spamy, prediction1))

acc_s.append(accuracy_score(spamy, prediction2))
pre_s.append(precision_score(spamy, prediction2))
rec_s.append(recall_score(spamy, prediction2))
f1_s.append(f1_score(spamy, prediction2))

acc_s.append(accuracy_score(spamy, prediction3))
pre_s.append(precision_score(spamy, prediction3))
rec_s.append(recall_score(spamy, prediction3))
f1_s.append(f1_score(spamy, prediction3))

data_set_name = ['GaussianNB', 'BernoulliNB', 'MultinomialNB']
result = pd.DataFrame({'精度': acc_s, '查准率': pre_s, '查全率': rec_s, 'F1': f1_s})
l = []
for i in range(result.shape[0]):
    l.append(data_set_name[i])
result.index = l
result.head(result.shape[0])

精度       查准率       查全率        F1
GaussianNB     0.821778  0.700444  0.956977  0.808858
BernoulliNB    0.883938  0.881336  0.815223  0.846991
MultinomialNB  0.786351  0.732363  0.721456  0.726869

###### 双击此处填写
算法|精度|查准率|查全率|F1值
-|-|-|-|-
GaussianNB | 0.821778 | 0.700444 | 0.956977 | 0.808858
BernoulliNB	| 0.883938 | 0.881336 | 0.815223 | 0.846991
MultinomialNB |	0.786351 | 0.732363 | 0.721456 | 0.726869

###### 双击此处填写分析

分析如下：